kymatio/scattering2d/backend/torch_backend.py

```diff
@@ -0,0 +1,328 @@
+
+    if not x.is_contiguous():
+        raise RuntimeError('Tensors must be contiguous!')
+
+    if direction == 'C2R':
+        output = torch.irfft(x, 2, normalized=False, onesided=False) * x.size(-2) * x.size(-3)
+    elif direction == 'C2C':
+        if inverse:
+            output = torch.ifft(x, 2, normalized=False) * x.size(-2) * x.size(-3)
+        else:
+            output = torch.fft(x, 2, normalized=False)
+
+    return output
+
+def cdgmm(A, B, inplace=False):
+    """Complex pointwise multiplication.
+    
+        Complex pointwise multiplication between (batched) tensor A and tensor B.
+
+        Parameters
+        ----------
+        A : tensor
+            A is a complex tensor of size (B, C, M, N, 2).
+        B : tensor
+            B is a complex tensor of size (M, N, 2) or real tensor of (M, N, 1).
+        inplace : boolean, optional
+            If set to True, all the operations are performed inplace.
+
+        Returns
+        -------
+        C : tensor
+            Output tensor of size (B, C, M, N, 2) such that:
+            C[b, c, m, n, :] = A[b, c, m, n, :] * B[m, n, :].
+
+    """
+    if not iscomplex(A):
+        raise TypeError('The input must be complex, indicated by a last '
+                        'dimension of size 2.')
+
+    if B.ndimension() != 3:
+        raise RuntimeError('The filter must be a 3-tensor, with a last '
+                           'dimension of size 1 or 2 to indicate it is real '
+                           'or complex, respectively.')
+
+    if not iscomplex(B) and not isreal(B):
+        raise TypeError('The filter must be complex or real, indicated by a '
+                        'last dimension of size 2 or 1, respectively.')
+
+    if A.size()[-3:-1] != B.size()[-3:-1]:
+        raise RuntimeError('The filters are not compatible for multiplication!')
+
+    if A.dtype is not B.dtype:
+        raise TypeError('A and B must be of the same dtype.')
+
+    if A.device.type != B.device.type:
+        raise TypeError('A and B must be both on GPU or both on CPU.')
+
+    if A.device.type == 'cuda':
+        if A.device.index != B.device.index:
+            raise TypeError('A and B must be on the same GPU!')
+
+    if isreal(B):
+        if inplace:
+            return A.mul_(B)
+        else:
+            return A * B
+    else:
+        C = A.new(A.size())
+
+        A_r = A[..., 0].contiguous().view(-1, A.size(-2)*A.size(-3))
+        A_i = A[..., 1].contiguous().view(-1, A.size(-2)*A.size(-3))
+
+        B_r = B[...,0].contiguous().view(B.size(-2)*B.size(-3)).unsqueeze(0).expand_as(A_i)
+        B_i = B[..., 1].contiguous().view(B.size(-2)*B.size(-3)).unsqueeze(0).expand_as(A_r)
+
+        C[..., 0].view(-1, C.size(-2)*C.size(-3))[:] = A_r * B_r - A_i * B_i
+        C[..., 1].view(-1, C.size(-2)*C.size(-3))[:] = A_r * B_i + A_i * B_r
+
+        return C if not inplace else A.copy_(C)
+
+def finalize(s0, s1, s2):
```

Comment:
Are we fine with this name (I'm fine, I believe one of us wasn't however)?

Reply:
(I'm fine too)

Reply:
@janden ?

Reply:
OK for `concatenate`

Reply:
:grin: